File: libs/node_packages/ng-spark/models/src/query-result.ts
import { TranslatedString } from './translated-string';
import { PersistentObject } from './persistent-object';
import { AS_DETAIL_BREADCRUMBS_KEY } from './as-detail-conversions';

/**
 * What a query returns: the column metadata once, then one lightweight row per result.
 *
 * Rows used to be full `PersistentObject`s, so every row repeated the complete attribute
 * metadata the client already held from `/spark/types` and never read off the row. More
 * importantly it conflated two things that are not the same: a **row is a projection**, a
 * **persistent object is a document**. Nothing here is entitled to act on a row — every
 * mutating path re-loads by id and re-applies security server-side.
 */
export interface QueryResult {
  columns: QueryColumn[];
  items: QueryResultItem[];
  /** Rows matching the query after filtering, before paging. */
  totalItems: number;
  skip: number;
  take: number;
  typeHints?: Record<string, string>;
  /**
   * Custom actions withheld for THIS result, by name. Absent or empty means every action the
   * caller has the right to is offered.
   *
   * The mirror of `PersistentObject.disabledActions`, set server-side via
   * `CustomQueryArgs.DisableActions(...)`. The action catalogue is per type and is never told what
   * an execution returned, so this is where a per-result answer travels.
   *
   * An affordance, not a permission -- the action handler still refuses on its own terms.
   */
  disabledActions?: string[];
}

/**
 * The subset of column metadata a cell needs in order to render.
 *
 * Both {@link QueryColumn} (a query result's column) and `EntityAttributeDefinition` (a detail
 * page's attribute) satisfy this, which is what lets one cell component serve the query grid and
 * the AsDetail sub-table without either side knowing about the other's shape.
 */
export interface SparkCellColumn {
  name: string;
  label?: TranslatedString;
  /** Help text rendered as an [i] tooltip in the column header. Absent when the model declares none. */
  description?: TranslatedString;
  dataType: string;
  isArray?: boolean;
  renderer?: string;
  rendererOptions?: Record<string, any>;
  referenceType?: string;
  lookupReferenceType?: string;
  asDetailType?: string;
}

/** One column of a query result, sent once rather than repeated on every row. */
export interface QueryColumn extends SparkCellColumn {
  order: number;
  /**
   * Whether the grid offers a sort affordance for this column. Resolved server-side from the query's
   * override then the attribute, so it arrives already decided.
   *
   * This replaced an `isSortable` that carried the server's AsDetail *drag-reorder* flag — a
   * different concept that was `false` for every scalar column while the grid hard-coded every
   * header sortable anyway. Neither side read it. The attribute-level `isSortable` on `EntityType`
   * is the drag-reorder flag and stays.
   *
   * Absent means sortable, matching `isVisible`: a server predating the field still draws arrows.
   */
  canSort?: boolean;
  /** Whether the grid draws a filter cell for this column. Absent means filterable. */
  canFilter?: boolean;
  /**
   * Whether this column's distinct values may be listed. When `false` and `canFilter` is true the
   * panel offers a free-text filter instead of a value list — listing a column's values is a much
   * stronger disclosure than filtering by one already known.
   */
  canListDistincts?: boolean;
  /**
   * Whether the grid draws this column. `false` means the row carries the value but no column is
   * rendered — for a renderer that needs a sibling value (a lock glyph beside a name) without
   * spending a column on it. `showedOn` decides what ships; this decides what is drawn.
   */
  isVisible?: boolean;
  /** The query backing a Reference column, by name — the client's option source. */
  query?: string;
}

/**
 * One row: an id, a display string, and a value per column.
 *
 * Deliberately too weak to act on — no attribute metadata, no `can` block, no etag, because none
 * of those can be trusted from a projection.
 */
export interface QueryResultItem {
  /** Never null and unique within a result; the server refuses anything else. */
  id: string;
  /** What to show when the row is named rather than tabulated (a reference picker's value). */
  breadcrumb?: string;
  values: QueryResultItemValue[];
  /** Presentation hints for the whole row. Keys arrive lower-cased. */
  typeHints?: Record<string, string>;
}

/** One cell. */
export interface QueryResultItemValue {
  /** The {@link QueryColumn.name} this value belongs to. */
  key: string;
  /** Typed JSON, not a string — the server converts on the way out. */
  value?: any;
  /** For a reference cell: the target document's id, so a link needs no second lookup. */
  objectId?: string;
  /** For a reference cell: the target's display string. */
  breadcrumb?: string;
  /** Resolved display labels per referenced id, for a multi-reference cell. */
  breadcrumbs?: Record<string, string | null>;
  typeHints?: Record<string, string>;
}

/**
 * Every shape a renderer's `item` can arrive in.
 *
 * Three, and they are genuinely different objects rather than variations of one: a query grid hands
 * a {@link QueryResultItem}, an AsDetail sub-table hands the flat record `nestedPoToDisplayRow`
 * built, and a detail or form host hands the `PersistentObject` itself. A renderer reused across a
 * grid and an AsDetail table therefore sees two of them, which is why {@link valueFor} reads all
 * three rather than making every app write the branch.
 */
export type SparkRow = QueryResultItem | PersistentObject | Record<string, unknown>;

/**
 * The value a row carries for a column, whatever shape the row is in — or `undefined` when it
 * carries none.
 *
 * ```ts
 * const isPrivate = valueFor(this.item(), 'IsPrivate')?.value === true;
 * ```
 *
 * Always returns the **cell** rather than the bare value, because a reference cell is worth as much
 * for its `objectId` as for its text, and a helper that discarded it would have to be replaced the
 * first time a renderer wanted to link somewhere. `?.value` is the cost, once per call site.
 *
 * ⚠️ A value only reaches a grid row if its attribute is on the **query surface**. To read a sibling
 * the grid does not draw, mark it `"showedOn": "Query", "isVisible": false` — shipped, not drawn. An
 * attribute marked `"showedOn": "PersistentObject"` is absent from a row by design, and this returns
 * `undefined` for it; that is the first thing to check when a sibling read comes back empty.
 */
export function valueFor(item: SparkRow | null | undefined, columnName: string): QueryResultItemValue | undefined {
  if (!item) return undefined;

  if (isQueryRow(item)) {
    return item.values.find(v => v.key === columnName);
  }

  if (isPersistentObject(item)) {
    const attribute = item.attributes.find(a => a.name === columnName);
    if (!attribute) return undefined;
    return {
      key: attribute.name,
      value: attribute.value,
      // A single reference's value IS the target's id, which is what the row shape carries
      // separately. Deriving it here keeps a renderer from having to know that.
      objectId: attribute.dataType === 'Reference' && !attribute.isArray && attribute.value != null
        ? String(attribute.value)
        : undefined,
      breadcrumb: attribute.breadcrumb,
      breadcrumbs: attribute.breadcrumbs,
    };
  }

  const record = item as Record<string, unknown>;
  if (!(columnName in record)) return undefined;

  const value = record[columnName];
  // The flat row's side channel is populated only for single references, so its presence for this
  // key is what identifies one — the record itself carries no dataType to ask.
  const breadcrumb = (record[AS_DETAIL_BREADCRUMBS_KEY] as Record<string, string> | undefined)?.[columnName];

  return {
    key: columnName,
    value,
    objectId: breadcrumb !== undefined && value != null ? String(value) : undefined,
    breadcrumb,
  };
}

/**
 * Whether this row is a query result item.
 *
 * Discriminated on the *elements*, not just on `Array.isArray(values)`: a flat AsDetail record may
 * legitimately have a column named `values` holding an array, and mistaking one for the other would
 * silently read the wrong thing.
 */
export function isQueryRow(row: SparkRow | null | undefined): row is QueryResultItem {
  const values = (row as QueryResultItem | undefined)?.values;
  return Array.isArray(values) && (values.length === 0 || typeof values[0]?.key === 'string');
}

/** Whether this row is a full persistent object. Discriminated on elements, as above. */
export function isPersistentObject(row: SparkRow | null | undefined): row is PersistentObject {
  const attributes = (row as PersistentObject | undefined)?.attributes;
  return Array.isArray(attributes) && (attributes.length === 0 || typeof attributes[0]?.name === 'string');
}

/**
 * One column's value filter (#431): the values a caller picked, or the values they excluded.
 *
 * **Values, never labels.** Each entry is the raw stored value — the same `value` the distinct
 * endpoint returned, never the `label` beside it. A reference column's values are document ids while
 * its labels are resolved breadcrumb text, and two rows can legitimately render the same text, so
 * matching on labels would be both wrong and ambiguous.
 *
 * `includes` and `excludes` are the two halves of the inverse toggle rather than a flag: the panel
 * moves values between them. `null` is a real, selectable value meaning "no value"; an empty or
 * absent array is the absent filter.
 */
export interface QueryColumnFilter {
  name: string;
  includes?: unknown[];
  excludes?: unknown[];
}

/**
 * One selectable entry in a column's filter panel.
 *
 * Structurally identical to the datatable's own `DistinctValue`, deliberately — it is what lets the
 * HTTP response be handed straight to `[distincts]` with no mapper in between.
 */
export interface DistinctValue {
  /** The raw stored value. `null` is a real entry meaning "no value", not an absent one. */
  value: unknown;
  /** What the user reads. Never matched on. */
  label: string;
}

/**
 * A column's distinct values.
 *
 * `remaining` is always empty from this server: the datatable snapshots the list when a selection is
 * first made and derives both buckets itself by diffing. The server answers "what matches now".
 *
 * `hasMore` reports that the cap truncated the list, and the panel re-queries only when it is set or
 * the search term widens — so a dishonest `false` strands a user typing past the cap.
 */
export interface DistinctValuesResult {
  matching: DistinctValue[];
  remaining: DistinctValue[];
  hasMore: boolean;
}

File: libs/spark/MintPlayer.Spark.Abstractions/QueryResult.cs
namespace MintPlayer.Spark.Abstractions;

/// <summary>
/// What a query returns: the column metadata once, then one lightweight row per result.
/// </summary>
/// <remarks>
/// Rows used to be full <see cref="PersistentObject"/>s, which meant every row carried a complete
/// copy of the attribute metadata — label, dataType, rules, renderer options, and for an AsDetail
/// attribute the whole nested object graph — that the client already held from
/// <c>GET /spark/types</c> and never read off the row. Beyond the payload, it conflated two things
/// that are not the same: a <b>row is a projection</b>, and a <b>persistent object is a document</b>.
/// <para>
/// The separation is what makes composed rows safe. A row was never claimed to be a document, so
/// nothing is entitled to treat a posted row id as verified: every mutating path re-materializes
/// from the id through the same load path a detail page uses, and re-applies security there. That
/// is why row ids can be treated as hostile input without an integrity token on the wire.
/// </para>
/// </remarks>
public sealed class QueryResult
{
    /// <summary>The column metadata for every row in <see cref="Items"/>, sent once.</summary>
    public required IReadOnlyList<QueryColumn> Columns { get; init; }

    public required IReadOnlyList<QueryResultItem> Items { get; init; }

    /// <summary>Rows matching the query after filtering, before paging.</summary>
    public required int TotalItems { get; init; }

    public required int Skip { get; init; }
    public required int Take { get; init; }

    /// <summary>Presentation hints for the result as a whole. See <see cref="QueryResultItem"/>.</summary>
    public IReadOnlyDictionary<string, string>? TypeHints { get; init; }
}

/// <summary>
/// One column of a query result — the metadata the client needs to draw a cell, hoisted out of the
/// per-row attributes it used to be repeated on.
/// </summary>
public sealed class QueryColumn
{
    public required string Name { get; init; }

    /// <summary>
    /// Whether the grid should draw this column. <see langword="false"/> means <b>ship the value,
    /// do not draw it</b> — the row carries it for a renderer to read, and the grid skips it.
    /// </summary>
    /// <remarks>
    /// A presentation decision, so it is carried rather than applied server-side. What belongs on
    /// the query surface at all is <c>showedOn</c>, which is the flag the sort allow-list is checked
    /// against too — one rule, one place. Filtering here as well would make an attribute both
    /// sortable and column-less, and would leave an app no way to give a renderer a sibling value
    /// short of giving it a column.
    /// </remarks>
    public bool IsVisible { get; init; } = true;

    public TranslatedString? Label { get; init; }
    /// <summary>Help text for the column header's [i] tooltip (#348); carried from the attribute.</summary>
    public TranslatedString? Description { get; init; }
    public string DataType { get; init; } = "string";
    public int Order { get; init; }
    public bool IsArray { get; init; }
    public bool IsSortable { get; init; }

    /// <summary>The query backing a Reference column, by name — the client's option source.</summary>
    public string? Query { get; init; }

    public string? ReferenceType { get; init; }
    public string? LookupReferenceType { get; init; }
    public string? AsDetailType { get; init; }

    /// <summary>A registered custom renderer's name, and whatever configuration it declared.</summary>
    public string? Renderer { get; init; }
    public Dictionary<string, object>? RendererOptions { get; init; }

    /// <summary>Presentation hints for every cell in this column. See <see cref="QueryResultItem"/>.</summary>
    public IReadOnlyDictionary<string, string>? TypeHints { get; init; }
}

/// <summary>
/// One row: an id, a display string, and a value per column.
/// </summary>
/// <remarks>
/// Deliberately too weak to act on. A row carries no attribute metadata, no <c>can</c> block and no
/// etag, because none of those can be trusted from a projection — a computed row has no document
/// behind it to re-judge. Anything that mutates re-loads by <see cref="Id"/> first.
/// <para>
/// <b>Type hints</b> are an open, string-keyed presentation side-channel, merged column → item →
/// value with later winning. There is no registry and no validation, which is the point: an
/// application adds its own keys without a framework change. Keys are lower-cased once, here, so a
/// client never has to try two spellings.
/// </para>
/// </remarks>
public sealed class QueryResultItem
{
    /// <summary>
    /// The row's identity, never null and unique within a result — enforced when the result is
    /// built, because a null or duplicate id silently collapses a grid rather than failing.
    /// </summary>
    public required string Id { get; init; }

    /// <summary>
    /// What to show when the row is named rather than tabulated — a reference picker's selected
    /// value, the first column's link text. Resolved server-side so those surfaces need no second
    /// fetch.
    /// </summary>
    public string? Breadcrumb { get; init; }

    public required IReadOnlyList<QueryResultItemValue> Values { get; init; }

    /// <summary>Presentation hints for the whole row (conditional colouring, an extra class).</summary>
    public IReadOnlyDictionary<string, string>? TypeHints { get; init; }
}

/// <summary>One cell.</summary>
public sealed class QueryResultItemValue
{
    /// <summary>The <see cref="QueryColumn.Name"/> this value belongs to.</summary>
    public required string Key { get; init; }

    /// <summary>
    /// The value, typed as JSON rather than stringified.
    /// </summary>
    /// <remarks>
    /// A deliberate divergence from the framework this design follows, which sends the whole wire as
    /// strings and re-types client-side from the column. Spark already converts to JSON-typed values
    /// on the way out, so stringifying would <em>add</em> machinery, lose date and number fidelity,
    /// and churn every renderer for nothing.
    /// </remarks>
    public object? Value { get; init; }

    /// <summary>For a reference cell: the target document's id, so a link needs no second lookup.</summary>
    public string? ObjectId { get; init; }

    /// <summary>For a reference cell: the target's display string. For an array reference, see
    /// <see cref="Breadcrumbs"/>.</summary>
    public string? Breadcrumb { get; init; }

    /// <summary>Resolved display labels per referenced id, for a multi-reference cell.</summary>
    public Dictionary<string, string?>? Breadcrumbs { get; init; }

    public IReadOnlyDictionary<string, string>? TypeHints { get; init; }
}
